# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [145]:
import pandas as pd
import numpy as np

In [146]:
id_names = np.arange(1,29,1)
list_id = [1,2,3,4,5,11,13,23]

In [147]:
df = pd.read_csv('horse_data.csv', names = list(id_names), na_values = '?')
df.info()
df_new = df.filter(items = list_id)
for i in list_id:
    pd.to_numeric(df_new[i])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   1       299 non-null    float64
 1   2       300 non-null    int64  
 2   3       300 non-null    int64  
 3   4       240 non-null    float64
 4   5       276 non-null    float64
 5   6       242 non-null    float64
 6   7       244 non-null    float64
 7   8       231 non-null    float64
 8   9       253 non-null    float64
 9   10      268 non-null    float64
 10  11      245 non-null    float64
 11  12      256 non-null    float64
 12  13      244 non-null    float64
 13  14      196 non-null    float64
 14  15      194 non-null    float64
 15  16      53 non-null     float64
 16  17      198 non-null    float64
 17  18      182 non-null    float64
 18  19      271 non-null    float64
 19  20      267 non-null    float64
 20  21      135 non-null    float64
 21  22      102 non-null    float64
 22  23

In [148]:
dict_names = {1:'surgery', 2:'Age', 3:'Hospital_Number', 4:'rectal_temperature', 
              5:'pulse', 11:'pain', 13:'abdominal_distension', 23:'outcome'}
df_new.rename(columns=dict_names, inplace=True)

In [149]:
describe= df_new.describe().T.reset_index()
dict_modes = {}
for v in dict_names.values():
    dict_modes[v] = df_new[v].mode()[0]
modes = pd.DataFrame.from_dict(dict_modes, orient='index').reset_index()
modes.rename(columns = {0:'mode'}, inplace = True)

In [150]:
df_new

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
0        2.0    1           530101                38.5   66.0   5.0   
1        1.0    1           534817                39.2   88.0   3.0   
2        2.0    1           530334                38.3   40.0   3.0   
3        1.0    9          5290409                39.1  164.0   2.0   
4        2.0    1           530255                37.3  104.0   NaN   
..       ...  ...              ...                 ...    ...   ...   
295      1.0    1           533886                 NaN  120.0   2.0   
296      2.0    1           527702                37.2   72.0   4.0   
297      1.0    1           529386                37.5   72.0   4.0   
298      1.0    1           530612                36.5  100.0   3.0   
299      1.0    1           534618                37.2   40.0   NaN   

     abdominal_distension  outcome  
0                     4.0      2.0  
1                     2.0      3.0  
2                     1.0      1.0  
3                     4.0      2.0  
4                     NaN      2.0  
..                    ...      ...  
295                   NaN      3.0  
296                   3.0      3.0  
297                   3.0      2.0  
298                   3.0      1.0  
299                   NaN      3.0  

[300 rows x 8 columns]

In [151]:
df_new.info()
describe.merge(modes, how='left')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   Hospital_Number       300 non-null    int64  
 3   rectal_temperature    240 non-null    float64
 4   pulse                 276 non-null    float64
 5   pain                  245 non-null    float64
 6   abdominal_distension  244 non-null    float64
 7   outcome               299 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 18.9 KB


index  count          mean           std       min  \
0               surgery  299.0  1.397993e+00  4.903046e-01       1.0   
1                   Age  300.0  1.640000e+00  2.173972e+00       1.0   
2       Hospital_Number  300.0  1.085889e+06  1.529801e+06  518476.0   
3    rectal_temperature  240.0  3.816792e+01  7.322887e-01      35.4   
4                 pulse  276.0  7.191304e+01  2.863056e+01      30.0   
5                  pain  245.0  2.951020e+00  1.307940e+00       1.0   
6  abdominal_distension  244.0  2.266393e+00  1.065131e+00       1.0   
7               outcome  299.0  1.551839e+00  7.371869e-01       1.0   

        25%       50%       75%        max      mode  
0       1.0       1.0       2.0        2.0       1.0  
1       1.0       1.0       1.0        9.0       1.0  
2  528904.0  530305.5  534727.5  5305629.0  527544.0  
3      37.8      38.2      38.5       40.8      38.0  
4      48.0      64.0      88.0      184.0      48.0  
5       2.0       3.0       4.0        5.0       3.0  
6       1.0       2.0       3.0        4.0       1.0  
7       1.0       1.0       2.0        3.0       1.0

## Краткие результаты исследования:
В качестве исходного материала нам достался датасет размера 300 строк на 28 столбцов.
Для более наглядного анализа отобрали 8 значимых столбцов:
1. Хирургия (surgery)
2. Возраст (Age)
3. Больничный ID (Hospital Number)
4. Ректальная температура (Rectal temperature)
5. Пульс (pulse)
6. Боли (pain)
7. Вздутие живота (abdominal distension)
8. Результат (outcome)


Соответственно все выводы ниже по урезанному датафрейму.
Как мы видим, что даже ключевые данные не обладают полнотой, мы видим всего лишь 2 столбца, в которых нет никаких пропусков (Возраст и ID лошади).



### Столбец Surgery(хирургия)
Описание:
хирургия? 
       -   1 = Да, была операция 
       -   2 = Лечение прошло без операции 
В данном столбце мб только 2 значения : 1 или 2 (описание выше).


Мы видим что у нас есть 1 пропуск в столбце, т.е. как минимум по одной лошади мы не обладаем информацией.
Выбросов по данному столбцу мы не видим, т.к. значения min и max у нас соответствуют нашим заявленным изначальным величинам (1 и 2).
Мода по данному столбцу = 1, т.е. мы можем сказать что в нашем датасете лошадей, у которых была операция, больше.
Т.к. величина категориальная, то остальные выводы неоднозначны.

### Столбец Age(возраст)
Возраст 
      -    1 = Взрослая лошадь 
      -    2 = Молодой (<6 месяцев) 
В данном столбце также только 2 значения.

Данный столбец обладает полнотой информации, т.е. заполнены все 300 строк, но имеет явный выброс(ошибку): значение max = 9 и значение std = 2.17 нам об этом говорят. Т.е. данный столбец нуждается в корректировке.
Мода = 1, т.е. в нашем датасете больше взрослых особей.
#### Работа с выбросами: 
надо заменить все значения 9 на значения 1.

In [152]:
upper_outliers = df_new[df_new.Age.between(8, 9, inclusive=True)].sort_values('Age')
upper_outliers

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
3        1.0    9          5290409                39.1  164.0   2.0   
262      2.0    9          5305629                38.6   88.0   NaN   
255      1.0    9          5294539                38.8  184.0   4.0   
252      1.0    9          5291719                38.8  124.0   2.0   
244      1.0    9          5291409                38.2  124.0   2.0   
229      1.0    9           534597                38.5  120.0   NaN   
227      1.0    9          5289419                 NaN    NaN   NaN   
212      1.0    9          5294369                38.9  120.0   3.0   
191      2.0    9          5287279                 NaN  100.0   4.0   
135      1.0    9          5287179                38.1  136.0   5.0   
109      1.0    9          5297379                38.4   84.0   3.0   
103      1.0    9          5292489                38.0  140.0   3.0   
82       1.0    9          5290759                38.1  100.0   3.0   
75       1.0    9           534092                39.7  100.0   2.0   
74       1.0    9          5292929                 NaN    NaN   NaN   
55       1.0    9          5282839                38.6  160.0   3.0   
41       2.0    9          5288249                39.0  150.0   NaN   
39       1.0    9          5277409                39.2  146.0   NaN   
23       1.0    9           534998                38.3  130.0   2.0   
16       1.0    9          5301219                 NaN  128.0   4.0   
13       2.0    9          5291329                38.0   92.0   1.0   
9        2.0    9          5299629                38.3   90.0   5.0   
275      1.0    9          5297159                38.8  150.0   5.0   
287      2.0    9          5305129                39.5   84.0   NaN   

     abdominal_distension  outcome  
3                     4.0      2.0  
262                   NaN      1.0  
255                   3.0      2.0  
252                   4.0      2.0  
244                   4.0      1.0  
229                   NaN      1.0  
227                   NaN      1.0  
212                   3.0      1.0  
191                   1.0      1.0  
135                   3.0      2.0  
109                   3.0      2.0  
103                   2.0      1.0  
82                    1.0      1.0  
75                    NaN      3.0  
74                    NaN      2.0  
55                    4.0      2.0  
41                    NaN      1.0  
39                    NaN      2.0  
23                    NaN      1.0  
16                    3.0      2.0  
13                    2.0      2.0  
9                     1.0      1.0  
275                   2.0      2.0  
287                   NaN      1.0

In [153]:
df_new.Age.unique()

array([1, 9], dtype=int64)

### Столбец Hospital number
По данному столбцу мы можем сказать лишь о его полноте, т.к. никаких других вводных условий не знаем.

### Столбец Rectal temperature
ректальная температура 
          - линейная 
          - в градусах Цельсия.
          - Повышенная температура может возникнуть из-за инфекции. 
          - температура может снижаться, когда животное находится в позднем шоке 
          - нормальная температура составляет 37,8 
          - этот параметр обычно изменяется по мере развития проблемы, 
               например. может вначале быть нормальным, затем стать повышенным из-за 
                   поражения, переходя обратно в нормальный диапазон, когда 
                   лошадь впадает в шок 
В данном столбце отсутствует 60 значений или 20%.
Глобальных выбросов мы не видим. Медиана, как и среднее колеблятся около значения 38,2. При условии что датасет у нас о лошадях с фактом наличия колики, то можем сказать, что такие значения нормальны.
Величина размаха и СКО между собой достаточно сильно разнятся, это говорит нам о том, что в массе своей значения температуры не столь вариативны, однако, встречаются некоторые аномалии.

#### Работа с выбросами

In [154]:
q1 = df_new.rectal_temperature.quantile(0.25)
q3 = df_new.rectal_temperature.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df_new[df_new.rectal_temperature.between(lower_bound, upper_bound, inclusive=True)].sort_values('rectal_temperature')
remove_outliers

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
113      1.0    1           527933                36.8   60.0   NaN   
277      2.0    1           528620                36.9   50.0   1.0   
70       2.0    1           528214                37.0   60.0   3.0   
292      1.0    1           530034                37.0   66.0   4.0   
190      1.0    1           528742                37.1   40.0   3.0   
..       ...  ...              ...                 ...    ...   ...   
165      1.0    1           534925                39.4  120.0   NaN   
231      2.0    1           530028                39.5   92.0   5.0   
287      2.0    9          5305129                39.5   84.0   NaN   
117      2.0    1          5290482                39.5    NaN   3.0   
162      2.0    1           535176                39.5   60.0   3.0   

     abdominal_distension  outcome  
113                   NaN      2.0  
277                   2.0      1.0  
70                    3.0      3.0  
292                   3.0      2.0  
190                   1.0      1.0  
..                    ...      ...  
165                   3.0      3.0  
231                   1.0      2.0  
287                   NaN      1.0  
117                   3.0      3.0  
162                   2.0      1.0  

[226 rows x 8 columns]

In [155]:
lower_outliers = df_new[df_new.rectal_temperature.between(df_new.rectal_temperature.min(), lower_bound, inclusive=True)].sort_values('rectal_temperature')
lower_outliers

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
44       1.0    1           535407                35.4  140.0   4.0   
141      2.0    1           522979                36.0   42.0   NaN   
238      2.0    1           528702                36.1   88.0   3.0   
80       1.0    1           527518                36.4   98.0   4.0   
118      1.0    1           533983                36.5   78.0   5.0   
298      1.0    1           530612                36.5  100.0   3.0   
251      2.0    1           527940                36.6   42.0   1.0   

     abdominal_distension  outcome  
44                    NaN      3.0  
141                   NaN      2.0  
238                   2.0      3.0  
80                    2.0      2.0  
118                   1.0      1.0  
298                   3.0      1.0  
251                   1.0      2.0

In [156]:
upper_outliers = df_new[df_new.rectal_temperature.between(upper_bound, df_new.rectal_temperature.max(), inclusive=True)].sort_values('rectal_temperature')
upper_outliers

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
99       2.0    1           530002                39.6  108.0   2.0   
75       1.0    9           534092                39.7  100.0   2.0   
20       1.0    1           530157                39.9   72.0   5.0   
281      2.0    1           530624                40.0   78.0   2.0   
54       2.0    1           529461                40.3  114.0   2.0   
91       2.0    1           529461                40.3  114.0   2.0   
259      1.0    1           527734                40.8   72.0   2.0   

     abdominal_distension  outcome  
99                    3.0      1.0  
75                    NaN      3.0  
20                    4.0      1.0  
281                   1.0      2.0  
54                    3.0      3.0  
91                    3.0      2.0  
259                   1.0      2.0

Предлагаю не выкидывать из выборки данные выбросы, т.к. аномалий мы не видим. Вполне вероятно что лошади во время лечения могли иметь как повышенные, так и пониженные температуры.

#### Столбец Pulse
пульс 
          - линейный 
          - частота пульса в ударах в минуту 
          - отражает состояние сердца: 30-40 нормально для взрослых 
          - редко бывает ниже нормы, хотя у спортивных лошадей
            может иметь частоту 20-25 
          - животные с болезненными поражениями или страдающие от шока кровообращения 
            могут иметь повышенную частоту сердечных сокращений 
            
Отсутствует 8% значений. Есть подозрение, что возможны выбросы. Я, конечно, не ветеринар, но пульс в 184 удара в минуту при норме в 30-40 весьма подозрителен. Разница между размахом и СКО говорит нам о вариативности значений. Мода, среднее и медиана также существенно различаются друг от друга.

#### Работа с выбросами

In [157]:
q1 = df_new.pulse.quantile(0.25)
q3 = df_new.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df_new[df_new.pulse.between(lower_bound, upper_bound, inclusive=True)].sort_values('pulse')
remove_outliers

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
130      1.0    1           528964                37.8   30.0   NaN   
232      1.0    1          5262543                38.5   30.0   NaN   
203      2.0    1           529685                37.2   36.0   2.0   
276      1.0    1           534787                38.0   36.0   4.0   
242      1.0    1           530354                 NaN   36.0   5.0   
..       ...  ...              ...                 ...    ...   ...   
148      1.0    1           529272                38.3  132.0   2.0   
135      1.0    9          5287179                38.1  136.0   5.0   
44       1.0    1           535407                35.4  140.0   4.0   
103      1.0    9          5292489                38.0  140.0   3.0   
39       1.0    9          5277409                39.2  146.0   NaN   

     abdominal_distension  outcome  
130                   NaN      2.0  
232                   NaN      1.0  
203                   1.0      1.0  
276                   2.0      3.0  
242                   3.0      2.0  
..                    ...      ...  
148                   2.0      1.0  
135                   3.0      2.0  
44                    NaN      3.0  
103                   2.0      1.0  
39                    NaN      2.0  

[271 rows x 8 columns]

In [158]:
lower_outliers = df_new[df_new.pulse.between(df_new.pulse.min(), lower_bound, inclusive=True)].sort_values('pulse')
lower_outliers

Empty DataFrame
Columns: [surgery, Age, Hospital_Number, rectal_temperature, pulse, pain, abdominal_distension, outcome]
Index: []

In [159]:
upper_outliers = df_new[df_new.pulse.between(upper_bound, df_new.pulse.max(), inclusive=True)].sort_values('pulse')
upper_outliers

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
41       2.0    9          5288249                39.0  150.0   NaN   
275      1.0    9          5297159                38.8  150.0   5.0   
55       1.0    9          5282839                38.6  160.0   3.0   
3        1.0    9          5290409                39.1  164.0   2.0   
255      1.0    9          5294539                38.8  184.0   4.0   

     abdominal_distension  outcome  
41                    NaN      1.0  
275                   2.0      2.0  
55                    4.0      2.0  
3                     4.0      2.0  
255                   3.0      2.0

С одной стороны, выбросы похожи на аномалии, но, если проаназировать по данным строчкам другие показатели, то данные выбросы можно интерпретировать как неудачные операции, в результате которых лошади не выжили. Предлагаю оставить.

#### Столбец pain
боль - субъективное суждение об уровне боли лошади 
          - возможные значения:
               1 = тревога, нет боли 
               2 = депрессия 
               3 = прерывистая легкая боль 
               4 = прерывистая сильная боль 
               5 = продолжительная сильная боль 
          - НЕ следует рассматривать как упорядоченную или дискретную переменную! 
          - В целом, чем более болезненно, тем больше вероятность того, что потребуется 
            хирургическое вмешательство 
          - предварительное лечение боли может в некоторой степени замаскировать уровень боли 

Отсутствует 18% значений. Выбросов не видим. Средняя, Мода и Медиана друг от друга не сильно отличаются.
Мода = 3, что говорит нам о том, что лошади переносили легкую прерывистую боль.

#### Столбец abdominal distension
вздутие живота 
          - ВАЖНЫЙ параметр. 
          - возможные значения 
               1 = нет 
               2 = незначительное 
               3 = умеренное 
               4 = тяжелое 
          - животное с вздутием живота, вероятно, будет болезненным и 
            у него снижена перистальтика кишечника. 
          - лошади с сильным вздутием живота, вероятно, потребуется 
            операция, чтобы просто сбросить давление 

Отсутсвует 19% значений. Выбросов не видим.
Среднее и Медиана отличаются друг от друга меньше, чем каждая по отдельности от Моды. Отсюда можем сделать вывод о том, что количество встречаемости достаточно невелико.

#### Столбец outcome
результат 
          - что в конечном итоге случилось с лошадью? 
          - возможные значения: 
               1 = жил 
               2 = умер 
               3 = был усыплен 

Мы видим что у нас есть 1 пропуск в столбце, т.е. как минимум по одной лошади мы не обладаем информацией. Аномалий или выбросов не видим, все в рамках указанных вводных. Значение СКО невелико, что, вкупе со значением Моды = 1 говорит нам о том, что 
основная часть лошадей жива, и это прекрасно!

### Пропуски


In [160]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   Hospital_Number       300 non-null    int64  
 3   rectal_temperature    240 non-null    float64
 4   pulse                 276 non-null    float64
 5   pain                  245 non-null    float64
 6   abdominal_distension  244 non-null    float64
 7   outcome               299 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 18.9 KB


In [161]:
for col in df_new.columns:
    pct_missing = df_new[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
Age - 0.0%
Hospital_Number - 0.0%
rectal_temperature - 20.0%
pulse - 8.0%
pain - 18.3%
abdominal_distension - 18.7%
outcome - 0.3%


In [162]:
#попробуем выкинуть строки DF, у которых отсутствует n из 9 значений
df_drop1 = df_new.dropna(thresh=1).info()
# ничего не изменилось

<class 'pandas.core.frame.DataFrame'>
Int64Index: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   Hospital_Number       300 non-null    int64  
 3   rectal_temperature    240 non-null    float64
 4   pulse                 276 non-null    float64
 5   pain                  245 non-null    float64
 6   abdominal_distension  244 non-null    float64
 7   outcome               299 non-null    float64
dtypes: float64(6), int64(2)
memory usage: 21.1 KB


In [163]:
df_new.loc[pd.isna(df_new.surgery),:]
#Уберем строчку с NaN в surgery и outcome, т.к. результата мы все равно не знаем.
df_new1 = df_new.loc[~pd.isna(df_new.surgery),:]
df_new1

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
0        2.0    1           530101                38.5   66.0   5.0   
1        1.0    1           534817                39.2   88.0   3.0   
2        2.0    1           530334                38.3   40.0   3.0   
3        1.0    9          5290409                39.1  164.0   2.0   
4        2.0    1           530255                37.3  104.0   NaN   
..       ...  ...              ...                 ...    ...   ...   
295      1.0    1           533886                 NaN  120.0   2.0   
296      2.0    1           527702                37.2   72.0   4.0   
297      1.0    1           529386                37.5   72.0   4.0   
298      1.0    1           530612                36.5  100.0   3.0   
299      1.0    1           534618                37.2   40.0   NaN   

     abdominal_distension  outcome  
0                     4.0      2.0  
1                     2.0      3.0  
2                     1.0      1.0  
3                     4.0      2.0  
4                     NaN      2.0  
..                    ...      ...  
295                   NaN      3.0  
296                   3.0      3.0  
297                   3.0      2.0  
298                   3.0      1.0  
299                   NaN      3.0  

[299 rows x 8 columns]

In [164]:
df_new1.loc[pd.isna(df_new.pulse),:]
# мы видим что в отсутствующих значениях по пульсу также отсутствует  ректальная температура. Попробуем оценить к-т корреляции.

surgery  Age  Hospital_Number  rectal_temperature  pulse  pain  \
5        2.0    1           528355                 NaN    NaN   2.0   
28       1.0    1          5279442                 NaN    NaN   NaN   
52       2.0    1           529483                 NaN    NaN   1.0   
56       1.0    1           528872                 NaN    NaN   NaN   
58       1.0    1           528298                 NaN    NaN   5.0   
74       1.0    9          5292929                 NaN    NaN   NaN   
78       1.0    1           530693                 NaN    NaN   5.0   
83       1.0    1          5279822                38.0    NaN   5.0   
93       2.0    1           530310                 NaN    NaN   5.0   
115      2.0    1           533723                 NaN    NaN   3.0   
117      2.0    1          5290482                39.5    NaN   3.0   
126      1.0    1           530384                38.7    NaN   5.0   
136      1.0    1           530693                 NaN    NaN   5.0   
151      1.0    1           529399                39.3    NaN   4.0   
159      2.0    1           528134                 NaN    NaN   2.0   
160      1.0    1           527916                 NaN    NaN   4.0   
174      2.0    1           518476                 NaN    NaN   4.0   
175      1.0    1           527929                 NaN    NaN   NaN   
182      1.0    1          5279822                38.0    NaN   5.0   
204      1.0    1           529528                39.2    NaN   4.0   
207      1.0    1           530297                 NaN    NaN   NaN   
216      2.0    1           528977                 NaN    NaN   NaN   
227      1.0    9          5289419                 NaN    NaN   NaN   
288      1.0    1           529428                 NaN    NaN   NaN   

     abdominal_distension  outcome  
5                     2.0      1.0  
28                    NaN      2.0  
52                    1.0      1.0  
56                    NaN      1.0  
58                    3.0      2.0  
74                    NaN      2.0  
78                    3.0      2.0  
83                    4.0      2.0  
93                    3.0      2.0  
115                   2.0      1.0  
117                   3.0      3.0  
126                   2.0      1.0  
136                   3.0      2.0  
151                   2.0      2.0  
159                   2.0      1.0  
160                   NaN      1.0  
174                   4.0      2.0  
175                   NaN      1.0  
182                   4.0      2.0  
204                   2.0      1.0  
207                   NaN      2.0  
216                   NaN      1.0  
227                   NaN      1.0  
288                   NaN      1.0

In [165]:
df_new1[['pulse', 'rectal_temperature']].corr()
# связь мала, так что, учитывая отсутствие выбросов,
#предлагаю именно эту часть значений Nan у пульса заполнить константой 9999, чтобы отдельно проанализировать.

pulse  rectal_temperature
pulse               1.000000            0.218868
rectal_temperature  0.218868            1.000000

In [166]:
df_new1.pulse.fillna('no_info', inplace = True)

C:\Users\Yarmullin\anaconda3\lib\site-packages\pandas\core\generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [167]:
df_new1.loc[pd.isna(df_new.rectal_temperature),:]
#учитывая что выбросов по температуре мы не имеем, а значения медианы, среднего и моды примерно одинаковы, 
#то заменим все отсутствующие данные на среднее.
df_new1.rectal_temperature.fillna(df_new.rectal_temperature.mean(), inplace=True)

In [168]:
df_new1.loc[pd.isna(df_new.abdominal_distension),:]
# аналогично пульсу - отсутствует почти всегда попарно с болями, поэтому заменим на 9000, чтобы позже проанализировать.
df_new1.abdominal_distension.fillna('no_info', inplace = True)

In [169]:
df_new1.loc[pd.isna(df_new.pain),:]
#на всякий случай тоже заменим на 9999, попарного сходства у них с вздутием много
df_new1.pain.fillna('no_info', inplace = True)

In [170]:
#заменим в столбце age значение 9 модальным значением.


In [171]:
df_new1.replace(9,1, inplace = True)

C:\Users\Yarmullin\anaconda3\lib\site-packages\pandas\core\frame.py:4172: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


In [172]:
df_new1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   299 non-null    int64  
 2   Hospital_Number       299 non-null    int64  
 3   rectal_temperature    299 non-null    float64
 4   pulse                 299 non-null    object 
 5   pain                  299 non-null    object 
 6   abdominal_distension  299 non-null    object 
 7   outcome               299 non-null    float64
dtypes: float64(3), int64(2), object(3)
memory usage: 21.0+ KB


In [176]:
df_new1

surgery  Age  Hospital_Number  rectal_temperature pulse     pain  \
0        2.0    1           530101           38.500000    66        5   
1        1.0    1           534817           39.200000    88        3   
2        2.0    1           530334           38.300000    40        3   
3        1.0    1          5290409           39.100000   164        2   
4        2.0    1           530255           37.300000   104  no_info   
..       ...  ...              ...                 ...   ...      ...   
295      1.0    1           533886           38.167917   120        2   
296      2.0    1           527702           37.200000    72        4   
297      1.0    1           529386           37.500000    72        4   
298      1.0    1           530612           36.500000   100        3   
299      1.0    1           534618           37.200000    40  no_info   

    abdominal_distension  outcome  
0                      4      2.0  
1                      2      3.0  
2                      1      1.0  
3                      4      2.0  
4                no_info      2.0  
..                   ...      ...  
295              no_info      3.0  
296                    3      3.0  
297                    3      2.0  
298                    3      1.0  
299              no_info      3.0  

[299 rows x 8 columns]